Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.mm

```diff
@@ -0,0 +1,601 @@
+
+    // Determine if we should mux a video or audio frame. We favor the audio frame first to satisfy webm muxing requirement:
+    // https://www.webmproject.org/docs/container/
+    // - Audio blocks that contain the video key frame's timecode SHOULD be in the same cluster as the video key frame block.
+    // - Audio blocks that have same absolute timecode as video blocks SHOULD be written before the video blocks.
+    bool takeVideo = !m_encodedVideoFrames.isEmpty() && (m_encodedAudioFrames.isEmpty() || m_encodedVideoFrames.first().timestamp < sampleTime(m_encodedAudioFrames.first()).microsecondsAs<int64_t>());
+
+    if (takeVideo) {
+        auto frame = m_encodedVideoFrames.takeFirst();
+        if (frame.isKeyFrame && frame.timestamp)
+            m_delegate->forceNewClusterOnNextFrame();
+        Seconds pts = Seconds::fromMicroseconds(frame.timestamp);
+        LOG(MediaStream, "muxNextFrame: writing video frame time:%f (previous:%f)", pts.value(), m_lastMuxedSampleTime.value());
+        ASSERT(pts >= m_lastMuxedSampleTime);
+        m_lastMuxedSampleTime = pts;
+
+        return m_delegate->addFrame({ frame.data.data(), frame.data.size() }, m_videoTrackIndex, m_lastMuxedSampleTime.nanosecondsAs<uint64_t>(), frame.isKeyFrame);
+    }
+
+    ASSERT(!m_encodedAudioFrames.isEmpty());
+    RetainPtr sampleBuffer = m_encodedAudioFrames.takeFirst();
+    auto buffer = PAL::CMSampleBufferGetDataBuffer(sampleBuffer.get());
+    ASSERT(PAL::CMBlockBufferIsRangeContiguous(buffer, 0, 0));
+    size_t srcSize = PAL::CMBlockBufferGetDataLength(buffer);
+    char* srcData = nullptr;
+    if (PAL::CMBlockBufferGetDataPointer(buffer, 0, nullptr, nullptr, &srcData) != noErr)
+        return false;
+
+    maybeForceNewCluster();
+
+    auto pts = sampleTime(sampleBuffer);
+    LOG(MediaStream, "muxNextFrame: writing audio frame time:%f (previous:%f)", pts.value(), m_lastMuxedSampleTime.value());
+    ASSERT(pts >= m_lastMuxedSampleTime);
+    m_lastMuxedSampleTime = pts;
+    return m_delegate->addFrame({ reinterpret_cast<const uint8_t*>(srcData), srcSize }, m_audioTrackIndex, m_lastMuxedSampleTime.nanosecondsAs<uint64_t>(), true);
+}
+
+void MediaRecorderPrivateWriterWebM::stopRecording()
+{
+    if (m_isStopping)
+        return;
+
+    LOG(MediaStream, "stopRecording");
+    m_isStopping = true;
+
+    if (m_audioCompressor)
+        RefPtr { m_audioCompressor }->finish();
+
+    flushPendingData()->whenSettled(MainThreadDispatcher::singleton(), [protectedThis = Ref { *this }, this] {
```

Comment:
Why not weakThis here, it seems the only place where we take a strong ref?

Reply:
we're stopping the encoder, we want the completionHandler to be called under all circumstances. Keeping the ref ensure everything will be completed nicely and things fully cleaned up